File: Ahorn/entities/FlagTempleGate.jl
module XaphanHelperFlagTempleGate

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/FlagTempleGate" FlagTempleGate(x::Integer, y::Integer,  width::Integer=8, height::Integer=48, flag::String="", spriteName::String="default", openOnHeartCollection::Bool=false, startOpen::Bool=false, horizontal::Bool=false, attachRight::Bool=false)

const placements = Ahorn.PlacementDict(
    "Flag Temple Gate (Xaphan Helper)" => Ahorn.EntityPlacement(
        FlagTempleGate,
        "point",
        Dict{String, Any}(
            "spriteName" => "default"
        )
    ),
)

textures = String["default", "mirror", "theo"]

Ahorn.editingOptions(entity::FlagTempleGate) = Dict{String, Any}(
    "spriteName" => textures
)

function Ahorn.selection(entity::FlagTempleGate)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 8))
    horizontal = get(entity.data, "horizontal", false)
    if horizontal
        return Ahorn.Rectangle(x, y - 3, height, 15)
    else
        return Ahorn.Rectangle(x - 3, y, 15, height)
    end
    
end

sprites = Dict{String, String}(
    "default" => "objects/door/TempleDoor00",
    "mirror" => "objects/door/TempleDoorB00",
    "theo" => "objects/door/TempleDoorC00"
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FlagTempleGate, room::Maple.Room)
    sprite = get(entity.data, "spriteName", "default")
    horizontal = get(entity.data, "horizontal", false)
    attachRight = get(entity.data, "attachRight", false)

    if haskey(sprites, sprite)
        if horizontal
            if attachRight
                Ahorn.drawSprite(ctx, sprites[sprite], 55, 21, rot=pi / 2)
            else
                Ahorn.drawSprite(ctx, sprites[sprite], 7, 36, rot=-pi / 2)
            end
        else
            Ahorn.drawSprite(ctx, sprites[sprite], 4, 24)
        end
    end
end

end

File: Loenn/entities/FlagTempleGate.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local FlagTempleGate = {}

FlagTempleGate.name = "XaphanHelper/FlagTempleGate"
FlagTempleGate.depth = -9000
FlagTempleGate.fieldOrder = {
    "x", "y", "flag", "spriteName", "horizontal", "attachRight", "startOpen", "openOnHeartCollection"
}
FlagTempleGate.ignoredFields = {
    "_name", "_id", "width", "height"
}
FlagTempleGate.fieldInformation = {
    spriteName = {
        options = {"default", "mirror", "theo"}
    }
}
FlagTempleGate.canResize = {false, false}
FlagTempleGate.placements = {
    name = "FlagTempleGate",
    data = {
        width = 8,
        height = 48,
        flag = "",
        startOpen = false,
        spriteName = "default",
        openOnHeartCollection = false,
        horizontal = false,
        attachRight = false
    }
}

function FlagTempleGate.sprite(room, entity)
    local horizontal = entity.horizontal or false
    local attachRight = entity.attachRight or false
    local directory = {}
    directory["default"] = "objects/door/TempleDoor00"
    directory["mirror"] = "objects/door/TempleDoorB00"
    directory["theo"] = "objects/door/TempleDoorC00"
    local texture = directory[entity.spriteName] or directory["mirror"]
    
    local sprite = drawableSprite.fromTexture(texture, entity)

    if horizontal then
        if attachRight then
            sprite.rotation = math.pi / 2
            sprite:addPosition(24, 4)
        else
            sprite.rotation = -math.pi / 2
            sprite:addPosition(24, 5)
        end
        
    else
        sprite.rotation = 0
        sprite:addPosition(4, 24)
    end

    return sprite
end

function FlagTempleGate.selection(room, entity)
    local horizontal = entity.horizontal or false
    local width = 8
    local height = 48
    if horizontal then
        width = 48
        height = 8
    end
    return utils.rectangle(entity.x, entity.y , width, height)
end

FlagTempleGate.offset = {4, 0}

return FlagTempleGate